src/mon/OSDMonitor.cc

```diff
@@ -3341,24 +3346,38 @@ epoch_t OSDMonitor::send_pg_creates(int osd, Connection *con, epoch_t next) cons
              << " : epoch " << epoch << " " << pgs.size() << " pgs" << dendl;
     last = epoch;
     for (auto& pg : pgs) {
-      if (!m)
-	m = new MOSDPGCreate(creating_pgs_epoch);
       // Need the create time from the monitor using its clock to set
       // last_scrub_stamp upon pg creation.
-      auto create = creating_pgs.pgs.find(pg);
+      auto create = creating_pgs.pgs.find(pg.pgid);
       assert(create != creating_pgs.pgs.end());
-      m->mkpg.emplace(pg, pg_create_t{create->second.first, pg, 0});
-      m->ctimes.emplace(pg, create->second.second);
+      if (old) {
+	if (!oldm) {
+	  oldm = new MOSDPGCreate(creating_pgs_epoch);
+	}
+	oldm->mkpg.emplace(pg.pgid,
+			   pg_create_t{create->second.first, pg.pgid, 0});
+	oldm->ctimes.emplace(pg.pgid, create->second.second);
+      } else {
+	if (!m) {
+	  m = new MOSDPGCreate2(creating_pgs_epoch);
+	}
+	m->pgs.emplace(pg, create->second);
+      }
       dout(20) << __func__ << " will create " << pg
 	       << " at " << create->second.first << dendl;
     }
   }
-  if (!m) {
+  if (!oldm && !m) {
     dout(20) << __func__ << " osd." << osd << " from " << next
              << " has nothing to send" << dendl;
     return next;
   }
-  con->send_message(m);
+  if (m) {
+    con->send_message(m);
+  }
+  if (oldm) {
```

Comment:
This is a little awkward, since we know you can't have both m and oldm. Make it an else-if, instead of sequential if blocks? Or just combine them down to one line with an inlined selection?

Reply:
fixed